### PR TITLE
Key value filter

### LINKED ISSUE #15: Filter keys/values based on filter results
Per @mapsam and @springmeyer 's thoughts:

In the current implementation we shave layers and whole features, but we do not (yet) drop unused properties from features that we keep. So if we keep a feature we currently decode all its properties and re-add all of its properties. If we only re-add'ed properties that are needed by a given style that would likely result in smaller, more efficient tiles for styles that only reference a few properties per style layer using vector tile sources with many properties.

https://github.com/mapbox/vtshaver/blob/da0b5bad9bcf08368f47cb97081cd2917bdfab0d/src/shave.cpp#L322

Starting to shave off unused properties could be a particularly good optimization for when vector tiles contain many language fields but a given style only needs to display few or none of them for a given layer.

Overall next steps to test this feature:

- [x] Extra parsing of the style by https://github.com/mapbox/vtshaver/blob/master/lib/styleToFilters.js to collect properties we need to keep based on keys referenced in style fields like `text-field` and any fields referenced in expressions if the styles uses expressions (https://blog.mapbox.com/announcing-expressions-in-gl-js-a72b55d0a6af)
- [x] Property whitelisting during shaving: Add the ability to pass down an array of extra "properties to keep" to the C++ shave function so that they are not stripped off.

### PR DESCRIPTION
**Pull request by zmofei**
Add the feature for vtshaver to shave the unused properties key and value.

Closes #15

### CODE REVIEW

File: src/filters.cpp [hunk 1]
```diff
@@ -157,9 +157,36 @@ NAN_METHOD(Filters::New) {
...
+                    }
+
+                    // NOTICE: If a layer is styled, but does not have a property, the property value will equal []
+                    // NOTICE: If a property is true, that means we need to keep all the properties
+                    filter_properties property;
+                    if (layer_properties->IsArray()) {
+                        // const auto propertyArray = layer_properties.As<v8::Array>();
+                        v8::Handle<v8::Array> propertyArray = v8::Handle<v8::Array>::Cast(layer_properties);
+                        uint32_t propertiesLength = propertyArray->Length();
+                        std::vector<std::string> values(propertiesLength);
```
Comment (springmeyer) [inline src/filters.cpp]: This is more efficient and safe:

```
std::vector<std::string> values;
values.reserve(propertiesLength);
for (uint32_t index = 0; index < propertiesLength; ++index) {
     Nan::Utf8String utf8_value(property_value);
     int utf8_len = utf8_value.length();
     if (utf8_len > 0) {
         values.emplace_back(*utf8_value, static_cast<std::size_t>(utf8_len));
     }
}
```

File: src/filters.hpp [hunk 2]
```diff
@@ -10,9 +10,13 @@
 class Filters : public Nan::ObjectWrap {
   public:
     using filter_value_type = mbgl::style::Filter;
+    using filter_properties = struct {
```
Comment (springmeyer) [inline src/filters.hpp]: I recommend:

 -  calling this `filter_properties_type`. 
  - making this a `std::pair` instead of a custom `struct` (more idiomatic in C++)
  - having the `type` be an enum for faster comparisons (rather than string)

File: src/filters.hpp [hunk 3]
```diff
@@ -23,9 +27,9 @@ class Filters : public Nan::ObjectWrap {
 
     static Nan::Persistent<v8::FunctionTemplate>& constructor();
 
-    void add_filter(filter_key_type&& key, filter_value_type&& filter, zoom_type minzoom, zoom_type maxzoom) {
+    void add_filter(filter_key_type&& key, filter_value_type&& filter, filter_properties properties, zoom_type minzoom, zoom_type maxzoom) {
```
Comment (springmeyer) [inline src/filters.hpp]: This should be passed as `filter_properties && properties` which means the object is movable in memory. Then you will need to use `std::move` to pass the variable to this function. See https://github.com/mapbox/cpp/blob/master/glossary.md#heap-allocation for more details on `std::move`.

File: src/filters.hpp [hunk 4]
```diff
@@ -23,9 +27,9 @@ class Filters : public Nan::ObjectWrap {
...
         // add a new key/value pair, with the value equaling a tuple 'filter_values_type' defined above
-        filters.emplace(key, std::make_tuple(std::move(filter), minzoom, maxzoom));
+        filters.emplace(key, std::make_tuple(std::move(filter), properties, minzoom, maxzoom));
```
Comment (springmeyer) [inline src/filters.hpp]: You'll need `std::move(properties)` here after you use `filter_properties && properties` above.

File: src/shave.cpp [hunk 5]
```diff
@@ -350,7 +351,26 @@ void filterFeatures(vtzero::tile_builder* finalvt,
                 feature_builder.set_id(feature.id());
             }
             feature_builder.set_geometry(feature.geometry());
+
+            std::string property_filter_type = property_filter.type;
+            std::vector<std::string> properties = property_filter.values;
+            auto keytable = layer.key_table();
```
Comment (springmeyer) [inline src/shave.cpp]: This should be:

```
            std::string const& property_filter_type = property_filter.type;
            std::vector<std::string> const& properties = property_filter.values;
            auto const& keytable = layer.key_table();
```

To avoid copies (potentially expensive) being made of those objects. In short, in Javascript every variable is a reference. In C++ you need to explicitly ask for a reference. The `&` means it is a reference, and therefore not copied.

File: src/shave.cpp [hunk 6]
```diff
@@ -350,7 +351,26 @@ void filterFeatures(vtzero::tile_builder* finalvt,
...
+
+            bool needAllProperties = property_filter_type == "all";
+            // if (property_filter_type != "all") {
             while (auto idxs = feature.next_property_indexes()) {
+                if (!needAllProperties) {
+                    // get the key only if we don't need all the properties;
+                    std::string key = keytable[idxs.key().value()].to_string();
```
Comment (springmeyer) [inline src/shave.cpp]: How about - for best performance - instead of comparing `std::string` here, we compare `uint32_t`? I think this could be accomplished by reversing the `properties` before the `next_property_indexes()` loop. You'd create a `property_idxs` vector that represents the indexs for all the property names. Then inside the `next_property_indexes()` you could see if the index value is in that `property_idxs` vector.

File: src/shave.cpp [hunk 7]
```diff
@@ -350,7 +351,26 @@ void filterFeatures(vtzero::tile_builder* finalvt,
                 feature_builder.set_id(feature.id());
             }
             feature_builder.set_geometry(feature.geometry());
+
+            std::string property_filter_type = property_filter.type;
+            std::vector<std::string> properties = property_filter.values;
+            auto keytable = layer.key_table();
+
+            bool needAllProperties = property_filter_type == "all";
```
Comment (springmeyer) [inline src/shave.cpp]: I think this calculation could be moved up in scope, for best efficiency, above `layer.for_each_feature`, since it does not change per feature.

File: src/filters.cpp [hunk 8]
```diff
@@ -157,9 +157,36 @@ NAN_METHOD(Filters::New) {
...
+                    const v8::Local<v8::Value> layer_properties = layer->Get(Nan::New("properties").ToLocalChecked());
+                    if (layer_properties->IsNull() || layer_properties->IsUndefined()) {
+                        Nan::ThrowError("Property-Filters is not properly constructed.");
+                        return;
+                    }
+
+                    // NOTICE: If a layer is styled, but does not have a property, the property value will equal []
+                    // NOTICE: If a property is true, that means we need to keep all the properties
+                    filter_properties property;
+                    if (layer_properties->IsArray()) {
```
[the remark this answers is not in the document]
Comment (zmofei) [reply src/filters.cpp]: yeah, In short words: The `filters` filert in styleTofilters.js use `true` for keep all. I've also used true to represent everything in order to be consistent.

------

more details: 
`filter_properties` means which property **to keep**,not which property **to remove**, so follow this idea, the empty means keep nothing, if we make it means keep all properties, that's a bit of a contradiction.

So, while I design this, I thought: In default, we don't need to keep any properties, but we need to collection the properties names since we start the traverse,so I use an array to store the properties which we need to keep. Only when we meat the key words `"properties"` which means we need to keep all the properties.  Follow the `filters` reules, I use `ture` to make this. 

so, true means every properties, `filter_properties` means which to keep, empty array means nothing to keep.

File: src/filters.cpp [hunk 9]
```diff
@@ -7,6 +7,7 @@
 // clang-format off
 #include <node/src/node_conversion.hpp>
 // clang-format on
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (zmofei) [reply src/filters.cpp]: 👍 yeah, I forget to remove it after I finished the debug work. will remove it.

File: src/filters.cpp [hunk 10]
```diff
@@ -157,9 +157,36 @@ NAN_METHOD(Filters::New) {
...
+                        return;
+                    }
+
+                    // NOTICE: If a layer is styled, but does not have a property, the property value will equal []
+                    // NOTICE: If a property is true, that means we need to keep all the properties
+                    filter_properties property;
+                    if (layer_properties->IsArray()) {
+                        // const auto propertyArray = layer_properties.As<v8::Array>();
+                        v8::Handle<v8::Array> propertyArray = v8::Handle<v8::Array>::Cast(layer_properties);
+                        uint32_t propertiesLength = propertyArray->Length();
```
[the remark this answers is not in the document]
Comment (zmofei) [reply src/filters.cpp]: Checked the [StackOverflow](https://stackoverflow.com/questions/14883896/why-is-stduint32-t-different-from-uint32-t), 
Learn this rule: `Always include what you use and don't rely on other headers including things for you`. will add the `std::`

File: src/filters.cpp (hunk 1 above)
Comment (zmofei) [reply src/filters.cpp]: Thanks, @springmeyer, I will make some change for this, but I still get some questions about this. 


1. Is this a risk: we reserve length to values, but we do not push anything into it if the value's length is 0.
Is that possible for traverse values and get the unpredictable value(if we reserve the length to 8, and we are not pushing eight values)?

-------
Questions need to research for @zmofei :
- [ ] `*v8::String::Utf8Value(property_value->ToString());` VS `Nan::Utf8String utf8_value`, how can we know the `utf8_value` method?
- [ ] `values.emplace_back(*utf8_value, static_cast<std::size_t>(utf8_len));` for the 2nd param what dose `static_cast<std::size_t>(utf8_len)` do and why we need pass a length to it?

File: src/filters.hpp (hunk 2 above)
Comment (zmofei) [reply src/filters.hpp]: done in the next commit:
* change the name
* use `std::pair`
* change the type into `enum`

File: src/filters.hpp (hunk 3 above)
Comment (zmofei) [reply src/filters.hpp]: 👍 got it(pass the memory into the next target), changed in next submit.

File: src/shave.cpp (hunk 5 above)
Comment (zmofei) [reply src/shave.cpp]: 👍 as same as the about one

File: src/filters.hpp (hunk 4 above)
Comment (zmofei) [reply src/filters.hpp]: 👍 got it(pass the memory into the next target), changed in next submit.

File: src/shave.cpp (hunk 5 above)
Comment (zmofei) [reply src/shave.cpp]: 👍 gotcha, & means we only pass the reference to the next step, not pass an copy of it.

File: src/shave.cpp (hunk 7 above)
Comment (zmofei) [reply src/shave.cpp]: 👍 Agree, To the extreme, from small details.

File: src/filters.cpp (hunk 1 above)
Comment (springmeyer) [reply src/filters.cpp]: > if we reserve the length to 8, and we are not pushing eight values

It is not a risk since the only thing `values.reserve` does is reserve memory. So, if we don't end up adding 8 the only problem is a little bit more memory that might be requested than used. However, a good rule of thumb is to only use `values.reserve` when we know that 100% of the time it will add up to 8 (or the `propertiesLength`). In this case I think it always will unless the developer has passed in an empty string in the properties. So, perhaps we should throw an error in this case?

Comment (springmeyer) [reply src/filters.cpp]: > how can we know the utf8_value method?

https://github.com/nodejs/nan/blob/master/doc/v8_misc.md#api_nan_utf8_string

> for the 2nd param what dose static_cast<std::size_t>(utf8_len) do and why we need pass a length to it?

This is tricky. So what is happening is that `emplace_back()` "forwards" the arguments to the constructor of the type inside the `vector`.  So the `Args&&` at https://en.cppreference.com/w/cpp/container/vector/emplace_back basically says "pass the arguments to the next function". That "next" function is the constructor for `std::string`. Now the `std::string` constructor has a lot of different signatures, but this one matches `string::string(charT const* s, size_type count)`: we are passing the `data` as the first arg and its `size` as the second (https://en.cppreference.com/w/cpp/string/basic_string/basic_string)

Comment (springmeyer) [reply src/filters.cpp]: In this case that is the only difference: when you know the length of the string then it makes sense to be explicit and use it. One other case where it could make a difference is if the ‘const char *’ data is not null terminated such that the length is required in order to know where the string ends. However I think in this case we don’t absolutely need to pass the length since I presume the data is null terminated.

File: src/filters.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (springmeyer) [reply src/filters.cpp]: 👍 Sounds good. Thanks for explaining.

File: src/filters.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (springmeyer) [reply src/filters.cpp]: 👍

File: src/filters.hpp (hunk 2 above)
Comment (springmeyer) [reply src/filters.hpp]: Looks great. Nice work.

File: src/shave.cpp (hunk 5 above)
Comment (springmeyer) [reply src/shave.cpp]: 👍

File: src/shave.cpp (hunk 6 above)
Comment (zmofei) [reply src/shave.cpp]: following the topic we talked in the channel,I use the new method to do this job:

```
auto const& keytable = layer.key_table();
std::vector<std::ptrdiff_t> props_by_index;
for (auto const& prop : properties) {
   auto itr = std::find(keytable.begin(), keytable.end(), prop);
   if (itr != keytable.end()) {
        props_by_index.emplace_back(std::distance(keytable.begin(), itr));
    }
}
```
and
```
if (std::find(props_by_index.begin(), props_by_index.end(), idxs.key()) == props_by_index.end()) {
    continue;
}
```

which can save about **5%** resource.

**Comment by zmofei** [comment]
@springmeyer Thanks for the feedback, I resolved all the point you comment. Now we have improved a lot in the performance, I make a test for run 1000000 times shave, and here are the average results:

**old**
```
node test/speed.js  94.43s user 2.74s system 376% cpu 25.824 total
```
**new**
```
node test/speed.js  61.48s user 2.59s system 374% cpu 17.098 total
```

which means we have about **35%** improve in the running time.

If you have time, please have a look at the new commit.

**Comment by springmeyer** [comment]
Excellent! In this case is “old” the master branch or is “old” this branch before PR review?

Headed to sleep now. Will review code in the morning!

**Comment by zmofei** [comment]
> is “old” the master branch or is “old” this branch before PR review?
The branch before PR review.

But this reminded me to make another test of the `master` branch(check out to the master, run `make` and copy the test/speed.js then run it). Run 3 times and the average result is:
```
node test/speed.js  80.12s user 4.57s system 301% cpu 28.077 total
```

The result is very interesting(If am not make some mistakes during the speed test), we add the new properties filter feature which means we do more jobs, but we still have some improvements in the performances(about **23%**[80.12s vs 61.48s]).

**Comment by springmeyer** [comment]
> The result is very interesting(If am not make some mistakes during the speed test), we add the new properties filter feature which means we do more jobs, but we still have some improvements in the performances(about 23%[80.12s vs 61.48s]).

Excellent that performance is slightly better in this branch than `master` @zmofei 🎉 I would hope/expect that given that encoding properties is expensive so the less we have to encode the more speed savings. However last I checked the overall top bottleneck is the `evaluate()` call, so all other work (more or less) will not impact performance as much as that call. So, given that a 23% perf gain is quite good. I see a similar slight gain when running `node bench/bench-batch.js` locally:

## master

```
$ node bench/bench-batch.js --iterations 500 --concurrency 1
Running benchmark...
Benchmark speed: 9259 runs/s (runs:15000 ms:1620 )
Note: pass --mem to track memory usage
Benchmark iterations: 500 concurrency: 1
Success: rate(9259.25925925926 ops/s) > min_rate(1000)
```

## This PR

```
$ node bench/bench-batch.js --iterations 500 --concurrency 1
Running benchmark...
Benchmark speed: 9615 runs/s (runs:15000 ms:1560 )
Note: pass --mem to track memory usage
Benchmark iterations: 500 concurrency: 1
Success: rate(9615.384615384615 ops/s) > min_rate(1000)
```

**Comment by springmeyer** [comment]
The next step on this PR I think is to improve % code coverage. We are seeing a slight drop in `filters.cpp`, which makes sense given there are new error conditions to handle. We should ensure that we have coverage of all the possible error conditions (sending invalid data for `properties`) so that we can ensure if invalid data is passed we never crash in c++.

**Comment by zmofei** [comment]
> We should ensure that we have coverage of all the possible error conditions

I've improved the test coverage, in the last commit which looks good I think.

image

By for some error catch, I think is not testable, for example:

### 1. case 1:

https://github.com/mapbox/vtshaver/blob/1e09582f4d230fda3c5af7f7512387c8b5aee4bc/src/filters.cpp#L69-L72

the `layers_val ` is come from v8 `GetPropertyNames`, which will always return an Array.
```C++
V8_DEPRECATE_SOON("Use maybe version", Local<Array> GetPropertyNames());
V8_WARN_UNUSED_RESULT MaybeLocal<Array> GetPropertyNames(Local<Context> context);
```

### 2. case 2:

https://github.com/mapbox/vtshaver/blob/1e09582f4d230fda3c5af7f7512387c8b5aee4bc/src/filters.cpp#L81-L84

The `layer_name_val` is come from an loop(`layers->Get(i)`) which we loop from an Array `for (std::uint32_t i = 0; i < layers->Length(); ++i) {` which means, it's impossible to get an unString value inside the loop(only if the C++  is broken??)

For this kind of code, how can we do the code coverage? @springmeyer do you have any suggestion? or we can put this kind of code into `// LCOV_EXCL_START & // LCOV_EXCL_STOP`?

**Comment by springmeyer** [comment]
👍 okay to use LCOV_EXCL if you think it cannot be covered. Just add a comment in the code about why.

**Comment by springmeyer** [comment]
@zmofei it looks like you might have already, but if not please pull in the latest changes from master now that https://github.com/mapbox/vtshaver/pull/16 is merged.

File: lib/styleToFilters.js [hunk 11]
```diff
@@ -18,7 +21,7 @@
 function styleToFilters(style) {
   var layers = {};
   // Store layers and filters used in style
-  if (style.layers) {
+  if (style && style.layers) {
```
Comment (springmeyer) [inline lib/styleToFilters.js]: What is the reason for this change @zmofei?

File: lib/styleToFilters.js [hunk 12]
```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
...
+    let value = propertyObj[key];
+    // TODO we still have outher situation:
+    // - special properties: `mapbox_clip_start`, `mapbox_clip_end`
+    if (typeof value === 'string') {
+      // if the value is string try to get property name from `xx{PropertyName}xx` like.
+      // the /{[^}]+}/ig return all the value like {xxx}
+      // eg 'a{hello}badfa' => ['{hello}']
+      // eg 'a{hello}ba{world}dfa' => ['{hello}','{world}']
+      let preProperties = value.match(/{[^}]+}/ig);
+      preProperties && preProperties.forEach(item => {
```
Comment (springmeyer) [inline lib/styleToFilters.js]: Let's is this style, which I think is more readable:

```
if (preProperties) {
    preProperties.forEach ....
}
```

File: lib/styleToFilters.js [hunk 13]
```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
...
+  })
+}
+
+
+function getPropertyFromExpression(exp, properties) {
+  // now we care about the expression like:
+  // ["get", string] not ["get", string, Object],
+  // ["has", string] not ["has", string, Object],
+  // ["properties"],
+  // ["feature-state", string]
```
Comment (springmeyer) [inline lib/styleToFilters.js]: Are these code comments necessary now that we have unit tests? If not I'd recommend removing them.

File: lib/styleToFilters.js [hunk 14]
```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
...
+  if (exp instanceof Array) {
+    switch (exp[0]) {
```
Comment (springmeyer) [inline lib/styleToFilters.js]: Could `exp` ever be empty or zero length?

File: lib/styleToFilters.js [hunk 15]
```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
...
+        if (typeof exp[1] === 'string' && !(exp[2] && typeof exp[2] === 'object')) {
+          properties.push(exp[1]);
+        }
+        break;
+      case 'feature-state':
+        properties.push(exp[1]);
+        break;
+      case 'properties':
+        properties.push(true);
+        break;
```
Comment (springmeyer) [inline lib/styleToFilters.js]: Should the `default` case (where non of those conditions match) be handled explicitly?

File: lib/styleToFilters.js [hunk 16]
```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
...
+    }
+
+    exp.forEach(sub => {
+      if (sub instanceof Array) {
```
Comment (springmeyer) [inline lib/styleToFilters.js]: We check for `instanceof Array` inside of `getPropertyFromExpression`. Maybe only do it one place?

File: lib/styleToFilters.js (hunk 11 above)
Comment (zmofei) [reply lib/styleToFilters.js]: Some time we may get the style from some where, it could be `null` or `underfined`. 
I know for this situation we have 2 options:
1. Through an Error to tell the user that this is illegal.
2. Do nothing.
Both of this 2 options are work I think, but the opt 1 may stop the rest of the code to run, And the opt 2 can tolerate `Non-fatal` error, it will not block the code to run and return empty `{}`, may friendly for the user and less code for use to write.

File: lib/styleToFilters.js (hunk 12 above)
Comment (zmofei) [reply lib/styleToFilters.js]: emmm, I am thinking for every possible types variable to go through `a && ...` and `if(a){ //...}`, but I don't find any difference between them in my mind.
But for writing "semantic" code, I'll change to `if`.

File: lib/styleToFilters.js (hunk 13 above)
Comment (zmofei) [reply lib/styleToFilters.js]: Yeah, unit tests can tell if all the condition are right. 
One of my concern is that I should let the code reader or the "6 month later me" can understand what we are doing here, and how many codition we need to deal with, below this comment are some trick code, with out the comment it's hard to understand, so I write this.
I think we should keep this, how do you think?

File: lib/styleToFilters.js (hunk 12 above)
Comment (springmeyer) [reply lib/styleToFilters.js]: Yes, just `semantic`. Both ways are 💯 valid, I'm just stating my preference for `if (a) { //... }`

File: lib/styleToFilters.js (hunk 13 above)
Comment (springmeyer) [reply lib/styleToFilters.js]: > I think we should keep this, how do you think?

Okay, 👍 to keeping. Can I ask that you make the comments more detailed then. It is hard for me to know what `["has", string] not ["has", string, Object],` means. I think it would be good to explain, in comments, with more detail.

File: lib/styleToFilters.js (hunk 14 above)
Comment (zmofei) [reply lib/styleToFilters.js]: Yes, good question, I should to judge the exp and I forget.
But while I review this line, I find it very interesting that, this time we are luck that we don't need to judge the exp, because:

1. if the exp is `null` or `undefined`, `exp instanceof Array` will return false.
2. if the exp is 0 length array, the `exp[0]` will be `undefined`, and in the next switch we only need to capture 4 conditions, it will not go into any condition, so we are fine.

File: lib/styleToFilters.js (hunk 11 above)
Comment (springmeyer) [reply lib/styleToFilters.js]: Great, thanks for explaining. I vote for `Throw an Error to tell the user that this is illegal.`. So code like:

```js
function styleToFilters(style) {
  var layers = {};
  // Store layers and filters used in style
  if (!style || !style.layers) {
      /// throw to avoid invalid usage
  }
}
```

While returning `{}` does seem like valid behavior I think it is preferable to have a way to present an error when this condition happens. Because, I can't think of a situation where `style` would be null and we would not, as developers, want to know about it. Can you?

File: lib/styleToFilters.js (hunk 15 above)
Comment (zmofei) [reply lib/styleToFilters.js]: humm, in this switch, we only need to consider 4 conditions, others we will do nothing. So are we need to add the empty default like:
```
default:
        break;
```

File: lib/styleToFilters.js (hunk 14 above)
Comment (springmeyer) [reply lib/styleToFilters.js]: Ah, I see. So, sounds like this code is okay given the way that Javascript works. In C++ `exp[0]` would crash since C++ does not (for performance reasons) automatically bounds check arrays, so my PR review 👀 always look for that. 

Okay, so this is 👍 to keep the same.

File: lib/styleToFilters.js (hunk 15 above)
Comment (springmeyer) [reply lib/styleToFilters.js]: 👍  on using `default` explicitly, if you agree. This is again, me making recommendations based on C++ guidelines and the conventions in JS may be difference, but I'm 👍 on explicitness due to http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-default. Feel free to disagree if you feel like being explicit is not idiomatic JS.

File: lib/styleToFilters.js (hunk 16 above)
Comment (zmofei) [reply lib/styleToFilters.js]: hummm, thinking....
`getPropertyFromExpression` designed to receive Array, even in the 1st line of the function we have `if (exp instanceof Array) {`, 
but I think we should prevent pass an NON-ARRAY param to this founction, so inside the function while before we start  recursive(call the `getPropertyFromExpression`) I write this `.
If (sub instanceof Array) {`, if we remove this, the whole program also work, but that way we may break `the function designed to receive Array, we should prevent pass an NON-ARRAY param to this founction,`. 
It seems becomes a philosophical question again. LOL.

File: lib/styleToFilters.js (hunk 15 above)
Comment (zmofei) [reply lib/styleToFilters.js]: Agree, I'll add this.

File: lib/styleToFilters.js (hunk 13 above)
Comment (zmofei) [reply lib/styleToFilters.js]: yeah, I'll add the more detail of the comment.

File: lib/styleToFilters.js (hunk 16 above)
Comment (springmeyer) [reply lib/styleToFilters.js]: Okay, if philosophical only, just make the decision yourself - I trust your judgement.

File: lib/styleToFilters.js (hunk 13 above)
Comment (springmeyer) [reply lib/styleToFilters.js]: 👌

File: lib/styleToFilters.js (hunk 16 above)
Comment (zmofei) [reply lib/styleToFilters.js]: Hmm, I choose the `if`.
`&&` it looks like a "hack trick", and the `if` is more common and better for others to read.

Comment (zmofei) [reply lib/styleToFilters.js]: oops, I got wrong content, I will keep the if check before recall the main function.

File: lib/styleToFilters.js (hunk 11 above)
Comment (zmofei) [reply lib/styleToFilters.js]: 👋 hey, Dane, I am trying to add the Erro, but after I added, I saw this test group in `styleToFilter.test.js#L13-L20`:

https://github.com/mapbox/vtshaver/blob/983b1e1a73429df327c7efbfd31bd49883892f22/test/styleToFilter.test.js#L13-L21

Thus, I think we should keep the original version without adding the Error. Otherwise, we need to change this test group, how do you think?

Comment (springmeyer) [reply lib/styleToFilters.js]: Oh, I see. I was not aware of those tests. Okay to keep without error then.

**Comment by springmeyer** [comment]
@zmofei I see a couple images that still need deleted from this PR `docs/image/howtodebug1.png` etc.

After that I think this is ready (given your findings at https://github.com/mapbox/vtshaver/issues/15#issuecomment-445765259) to:

 - [ ] merge
 - [ ] tag a formal release
 - [ ] publish binaries
 - [ ] publish

Feel free to do this today, or whenever you have time.

**Comment by springmeyer** [comment]
Note: @zmofei when you merge please choose "squash and merge" so that your change lands in master as a single commit.
